packages/langium/src/utils/caching.ts

```diff
@@ -0,0 +1,163 @@
+/******************************************************************************
+ * Copyright 2023 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import type { Disposable } from 'vscode-languageserver';
+import type { URI } from 'vscode-uri';
+import type { LangiumSharedServices } from '../services';
+
+export abstract class DisposableCache implements Disposable {
+
+    protected toDispose: Disposable;
+    protected isDisposed = false;
+
+    constructor(disposable: Disposable) {
```

Comment:
Proposal: add a method `onDispose` that adds a disposable to an array to separate this from construction.